Cosmo: validating and setting Parameters

nstarman opened the pull request:
Parameter value validation and allowing parameters to be "fixed"/derived, e.g. setting ``Ode0`` in ``FlatFLRWMixin``.

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

pep8speaks commented:
Hello @nstarman :wave:! Thanks for opening this pull request, we are very grateful for your contribution! I'm a friendly :robot: that checks for style issues in this pull request, since this project follows the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guidelines. I've listed some small issues I found below, but please don't hesitate to ask if any of them are unclear!


* In the file [`astropy/cosmology/core.py`](https://github.com/astropy/astropy/blob/5e3fe9a5a7af81562dcb31173fec6613b782d533/astropy/cosmology/core.py):

> [Line 239:1](https://github.com/astropy/astropy/blob/5e3fe9a5a7af81562dcb31173fec6613b782d533/astropy/cosmology/core.py#L239): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace



* In the file [`astropy/cosmology/tests/test_core.py`](https://github.com/astropy/astropy/blob/5e3fe9a5a7af81562dcb31173fec6613b782d533/astropy/cosmology/tests/test_core.py):

> [Line 175:64](https://github.com/astropy/astropy/blob/5e3fe9a5a7af81562dcb31173fec6613b782d533/astropy/cosmology/tests/test_core.py#L175): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

mhvk commented on astropy/cosmology/flrw.py:
```diff
@@ -1453,17 +1451,15 @@ class FlatFLRWMixin(FlatCosmologyMixin):
     def __init__(self, *args, **kw):
         super().__init__(*args, **kw)
         # Do some twiddling after the fact to get flatness
         self._Ode0 = 1.0 - self._Om0 - self._Ogamma0 - self._Onu0
```
My tendency would still be to go through `__set__`, i.e., `self.Ode0 = ...`

nstarman replied:
I would, but ``Ode0`` has already been set once. So only the private attribute is writeable.

mhvk replied:
Is that possible for people to put in `Ode0`? And then it gets overwritten? That seems a bit odd...

nstarman replied:
Sorry, I don't think I'm following this...

If you mean passing ``Ode0`` as an argument, then all the flat classes like `FlatLambdaCDM` don't take it as an argument.
I do agree that ``FlatFLRWMixin`` should explicitly prevent this!

nstarman replied:
I'm adding to FlatFLRWMixin

```python
def __init_subclass__(cls):
        super().__init_subclass__()
        if "Ode0" in cls._init_signature.parameters:
            raise TypeError("subclasses of `FlatFLRWMixin` cannot have `Ode0` in `__init__`")
```

nstarman replied:
class-level protection against `Ode0`.

mhvk replied:
I meant how is it possible for `Ode0` to already have been set here once at this place?

nstarman replied:
E.g. ``FlatLambdaCDM.__init__()``has 
```
super().__init__(H0=H0, Om0=Om0, Ode0=0.0, Tcmb0=Tcmb0, Neff=Neff,
                         m_nu=m_nu, Ob0=Ob0, name=name, meta=meta)
```
And then ``FlatFLRWMixin`` modifies the value.

It used to be
```
super().__init__(H0=H0, Om0=Om0, Ode0=0.0, Tcmb0=Tcmb0, Neff=Neff,
                         m_nu=m_nu, Ob0=Ob0, name=name, meta=meta)
self._Ode0 = ....
```

mhvk replied:
It seems that there should be a slightly better way... But no reason to hold this up for that!

nstarman replied:
Agreed! I think I have an idea for a solution. Going to require a small refactor of ``Parameter`` getters, so I'll get it in before the v5.0 lock!

mhvk commented on astropy/cosmology/core.py:
```diff
@@ -60,6 +64,19 @@ class Parameter:
         of the containing Cosmology.
         See https://docs.python.org/3/library/string.html#formatspec
 
+    fixed : bool (optional, keyword-only)
```
To avoid the confusion with model parameters, should this have a different name? E.g., `independent=False` or `derived=True`?

nstarman replied:
Hmm. A good point. I'll look at some Cosmology textbooks to see if there's a better nomenclature.

mhvk commented on astropy/cosmology/flrw.py:
```diff
@@ -2262,7 +2257,7 @@ def __init__(self, H0, Om0, Ode0, w0=-1.0, Tcmb0=0.0*u.K, Neff=3.04,
         super().__init__(H0=H0, Om0=Om0, Ode0=Ode0, Tcmb0=Tcmb0, Neff=Neff,
                          m_nu=m_nu, Ob0=Ob0, name=name, meta=meta)
-        self._w0 = float(w0)
+        self.w0 = float(w0)
```
In principle, this could be done with `Parameter(..., fvalidate=float, ...)` - which perhaps suggests that it should be `validate` rather than `fvalidate`...

nstarman replied:
Definitely yes to the first part.
I think ``fvalidate`` is good because its extending the ``fget/fset/fdel`` standard from property. Also ``Parameter.validate()`` calls ``fvalidate``, so it could be a little confusing to have an input match a function where they purposefully have different signatures.

mhvk commented on astropy/cosmology/flrw.py:
```diff
@@ -2877,9 +2872,9 @@ def __init__(self, H0, Om0, Ode0, wp=-1.0, wa=0.0, zp=0.0, Tcmb0=0.0*u.K,
-        self._zp = zp << self.__class__.zp.unit
+        self.wp = float(wp)
+        self.wa = float(wa)
+        self.zp = zp
```
Here, could one use `fvalidate=aszarr`?

nstarman replied:
Yes! that would be great. When https://github.com/astropy/astropy/pull/12169 is in ``fvalidate=aszarr`` is definitely the way to go.

nstarman replied:
Haha. Mispoke. That strips the units from ``zp``. I like the units!
I did speed up some internal calculations using ``self._zp.value`` instead of ``self._zp`` which always works b/c ``self._zp`` is always in units of redshift.

mhvk commented on astropy/cosmology/tests/test_core.py:
```diff
@@ -31,16 +31,22 @@ class Example1(Cosmology):
         class Example2(Example1):
             def __init__(self, param=15 * u.m):
-                self._param = param.to_value(u.km)
+                self._param = self.__class__.param.validate(self, param)
```
Becoming more and more convinced one should not limit `__set__`...

nstarman replied:
Agreed on allowing both ``getter`` and ``__set__``.

mhvk reviewed (commented):
I like how this is looking, but looking at some of the after-setting tests that remain, I wonder whether it might make things easier if one had two options:
1. A *verifier*, that just returns `True` if a value is OK, or `False` otherwise (with the default `__set__` raising on `False`); could then do things like `verifier=lambda x: x>=0.`  That said, in-line I'm suggesting `fvalidate=float`, so perhaps that has its uses too...
2. A *setter* where the user simply is in full control (but with the once-only still enforced by `__set__`). The advantage is that this is expected to exist from things like `(lazy)property`.

Mostly bringing this up since it would be good to get this right.

mhvk commented:
p.s. I'm wondering a bit whether it would make sense to in fact inherit from `lazyproperty` and add the validation (and override `__set__`, but still using `super()`). One nice bit with `lazyproperty` is that it was made thread-safe...

nstarman commented on astropy/cosmology/core.py:
```diff
@@ -81,15 +98,15 @@ class Parameter:
         ...     def __init__(self, param=15):
         ...         super().__init__(70, 0.3, 0.7)
-        ...         self._param = param << self.__class__.param.unit
+        ...         self._param = self.__class__.param.validate(self, param)
```
The more I look at it the more I agree. People defining a custom cosmology class are the type to double check some documentation or at least look at the code for examples. I'll put in appropriate hazard signs that custom getters can obviate the validation process

nstarman commented on astropy/cosmology/core.py:
```diff
@@ -106,18 +123,27 @@ class Parameter:
     """
 
-    def __init__(self, fget=None, doc=None, *, unit=None, equivalencies=[], fmt=".3g"):
+    def __init__(self, fget=None, fvalidate=None, doc=None, *,
```
A good point. ``fget`` is the important one and kwargs can be turned into optional args and maintain backward compatibility.

nstarman commented on astropy/cosmology/core.py:
```diff
@@ -106,18 +123,27 @@ class Parameter:
     """
 
-    def __init__(self, fget=None, doc=None, *, unit=None, equivalencies=[], fmt=".3g"):
+    def __init__(self, fget=None, fvalidate=None, doc=None, *,
```
Oops. Same comment. Resolving here.

nstarman commented on astropy/cosmology/core.py:
```diff
@@ -149,44 +175,127 @@ def format_spec(self):
+    # -------------------------------------------
+    # 'property' descriptor overrides
+
+    @property
```
Quite possibly. It's unlikely someone would mess with the attribute even if not protected, but modifying ``._fget``  is quite easy if one really needs to change ``fget``. But Parameter, like Cosmology, should be 'immutable'.

mhvk replied:
I think a `property.fget` really should be immutable too - it is only accessible via the class and it seems obvious one should not change it. So it would seem to add complexity and lose speed for very little benefit.

nstarman replied:
Perhaps ``Parameter`` should just inherit from `property`. Then things like ``fget`` would be automatically immutable.
``setter`` and ``deleter`` can just raise errors. 

```
class A:
    @property
    def attr(self):
        print(2)
a = A()
A.attr.fget = lambda self: print(3)  # fails
```

mhvk replied:
Indeed, that means circling back a bit, but it does make sense...

nstarman commented on astropy/cosmology/core.py:
```diff
@@ -149,44 +175,127 @@ def format_spec(self):
+        # raise error if setting 2nd time or has custom getter
+        # TODO! custom getter is currently a strong limitation on `validate`
+        #       because can't know if getter draws from ``_attr_name_private``
+        if hasattr(obj, self._attr_name_private) or self.fget is not None:
```
> Note: another option would be to have a flag on the Cosmology instance on whether one is still in "initialization mode". But what you have here feels cleaner.

Agreed. I tried that. It got hacky.

nstarman replied:
> Yet another option: lazyproperty stores the attribute under its own name in obj.__dict__ - since for a descriptor that is known to be safe. Would that make sense here too?

Let me time it all out again. I recall this implementation ended up being slow. Almost double the time.

nstarman commented on astropy/cosmology/flrw.py:
```diff
@@ -120,37 +120,36 @@ class FLRW(Cosmology):
         if Ob0 is not None:
-            self._Ob0 = float(Ob0)
+            self.Ob0 = float(Ob0)
             if self._Ob0 < 0.0:
```
Great! yeah. I'm moving a lot of ``__init__`` into validators for the parameters.

nstarman commented on astropy/cosmology/core.py:
```diff
@@ -230,6 +339,7 @@ class Cosmology(metaclass=abc.ABCMeta):
 
     # Parameters
     __parameters__ = ()
+    __all_parameters__ = ()
```
It would be nice to have only attribute tracking the parameters, but filtering for ``fixed=False`` every time we need ``__parameters__`` seems suboptimal. Maybe there's a good way around that. A ChainMap?

nstarman commented on astropy/cosmology/core.py:
```diff
@@ -149,44 +175,127 @@ def format_spec(self):
 
-    def getter(self, fget):
-        """Make new Parameter with custom ``fget``.
+    @property
```
The code internally uses ``_fvalidate`` to maintain speed. The immutability is to remain consistent with `property`. I don't particularly care for the indirectness, but prefer it to mutability. I agree most users are unlikely to use `Parameter`, but for those writing a custom Cosmology, I want to make the process nigh impossible to mess up. The tension between custom ``getter`` and the set-once is one such issue I plan to resolve.

nstarman commented on astropy/cosmology/core.py:
```diff
@@ -149,44 +175,127 @@ def format_spec(self):
+        return self.fvalidate(obj, self, value)
+
+    @staticmethod
+    def _default_validator(obj, param, value):
```
Agreed! I styled the validators after ``modeling.Parameter``, but I think this solution might be less verbose and more intuitive to those familiar with `property`. Might be worth pushing a refactor to ``modeling``. It's dev-side only. no API change.

nstarman commented on astropy/cosmology/flrw.py:
```diff
@@ -180,8 +179,8 @@ def __init__(self, H0, Om0, Ode0, Tcmb0=0.0*u.K, Neff=3.04, m_nu=0.0*u.eV,
-            with u.add_enabled_equivalencies(cls.m_nu.equivalencies):
-                m_nu = m_nu << cls.m_nu.unit
+            # validate value to Quantity using equivalencies
+            m_nu = self.__class__.m_nu.validate(self, m_nu)
```
It does simplify nicely!

nstarman commented:
Ok. Reverted back to validate. got rid of property and simplified accordingly. Hopefully this one's the one.
I definitely think this PR is stronger for the numerous back and forth.
I think I see some ways to improve further, but those are fine for v5.1.
Thanks @mhvk for all the iterations!

nstarman commented:
Rebased to fix merge conflict.

nstarman commented on astropy/cosmology/__init__.py:
```diff
@@ -8,16 +8,17 @@
 __all__ = (core.__all__ + flrw.__all__       # cosmology classes
            + realizations.__all__            # instances thereof
-           + funcs.__all__ + utils.__all__)  # utils
+           + funcs.__all__ + parameter.__all__ + utils.__all__)  # utils
```
Done!

nstarman commented on astropy/cosmology/parameter.py:
```diff
@@ -0,0 +1,326 @@
+        Function to get the value from instances of the cosmology class.
+        If None (default) returns the corresponding private attribute.
+        Often not set here, but as a decorator with ``getter``.
+    fset : callable[[object, object, Any], Any] or {'default', 'float'}, optional
```
Done!

astrofrog commented:
I am re-milestoning to v5.1 as we've now branched v5.0.x

nstarman commented on astropy/cosmology/parameter.py:
```diff
@@ -0,0 +1,326 @@
+_COSMOLOGY_CLASSES = dict()
+
+
+class Parameter(property):
```
Done!

nstarman commented on astropy/cosmology/parameter.py:
```diff
@@ -0,0 +1,326 @@
+            raise AttributeError("can't set attribute")
+
+        # validate value, generally setting units if present
+        value = self.set(cosmology, value)
```
Done!

nstarman commented on astropy/cosmology/parameter.py:
```diff
@@ -0,0 +1,326 @@
+        self._fmt = str(fmt)
+        self._derived = derived
+
+        # nested descriptor decorator compatibility
```
Removed.

mhvk commented on astropy/cosmology/io/tests/test_ecsv.py:
```diff
@@ -30,6 +31,14 @@ class ReadWriteECSVTestMixin(IOTestMixinBase):
     See ``TestCosmology`` for an example.
     """
 
+    @pytest.fixture
```
Is this still needed with your most recent addition to just this while reading?

nstarman replied:
I think it is. For a bare ``QTable.read()``, e.g. in ``test_tofrom_ecsv_instance``.

mhvk replied:
Ah, makes sense. I think there is a lot to be said for having the units enabled automatically on `astropy.cosmology` import, but definitely for another PR.

mhvk commented on astropy/cosmology/parameter.py:
```diff
@@ -0,0 +1,265 @@
+        Function to get the value from instances of the cosmology class.
+        If None (default) returns the corresponding private attribute.
+        Often not set here, but as a decorator with ``getter``.
+    fvalidate : callable[[object, object, Any], Any] or {'default', 'float'}, optional
```
List of pre-defined validators is incomplete... My tendency would be to just write `str` here.

nstarman replied:
SGTM. I'm adding a property ``registered_validators(self): return self._registry_validators.keys()`` so it's easy to discover what's a valid str.

nstarman replied:
made it a classproperty

nstarman replied:
Done!

mhvk commented on astropy/cosmology/parameter.py:
```diff
@@ -0,0 +1,265 @@
+
+    _registry_validators = {}
+
+    def __init__(self, fget=None, fvalidate="default", doc=None, *,
```
I don't think `fget` is used at all anymore (and if it is, would not do anything), so remove!?

nstarman replied:
Indeed!

nstarman replied:
That was a big oversight. Thanks for catching!

nstarman replied:
Done!

mhvk commented on astropy/cosmology/parameter.py:
```diff
@@ -0,0 +1,265 @@
+    """Parameter value validator where value is a positive float."""
+    value = _validate_to_float(cosmology, param, value)
+    if value < 0.0:
+        raise ValueError(f"{param.name} can not be negative.")
```
Total nitpick: "can not" -> "cannot"

nstarman replied:
SGTM. I think that's legacy, but I'm changing the intro to ``{param.name}``, so it's fine to change other things.

nstarman replied:
Done!

mhvk commented on astropy/cosmology/flrw.py:
```diff
@@ -262,6 +246,12 @@ def __init__(self, H0, Om0, Ode0, Tcmb0=0.0*u.K, Neff=3.04, m_nu=0.0*u.eV,
 
+        # now set m_nu Parameter (Note the value is calculated and not
+        # influenced by this value)
+        self.m_nu = m_nu
```
This part still feels slightly odd. Somehow it seems cleaner to just do the calculation and set the private property. But no big deal!

nstarman replied:
I could just do ``self.m_nu = ... ``  With the Ellipsis it feels less confusing.

nstarman replied:
Done!

mhvk reviewed (commented):
Looks good now! A few small comments, mostly that I think one more piece can be removed.

p.s. My tendency would be to ask permission to still get this in 5.0, so the new `Parameter` does not have to be changed.

astrofrog commented:
I am ok with backporting this as it fixes an API being introduced in 5.0. However please make sure the CI is all green before merging, there have been too many CI breakages in the last week 😅. The CircleCI builds might need restarting.

mhvk commented on astropy/cosmology/flrw.py:
```diff
@@ -114,52 +115,33 @@ class FLRW(Cosmology):
-        if not self._H0.isscalar:
-            raise ValueError("H0 is a non-scalar quantity")
+        # Derived quantities
+        self._Odm0 = None if Ob0 is None else (self._Om0 - self._Ob0)
```
This way of doing derived parameters seems nicer than the detailed calculation done below.

mhvk commented on astropy/cosmology/flrw.py:
```diff
@@ -185,10 +170,10 @@ def __init__(self, H0, Om0, Ode0, Tcmb0=0.0*u.K, Neff=3.04, m_nu=0.0*u.eV,
-                m_nu = m_nu << cls.m_nu.unit
+            self._neff_per_nu = self._Neff / self._nneutrinos
 
             # Now, figure out if we have massive neutrinos to deal with,
```
I'm still a bit confused about the logic here -- could all this logic be moved to the validator, thus avoiding the assignment at the end? Alternatively, might it make more sense to *use* the simple validator above and just update the private attribute at the end?

nstarman replied:
I think the latter is the better move for now. In this PR I refrained from assigning any attributes in the validator because I'm still unsure if that's a good code pattern.
E.g. 

```python
@H0.validator
def H0(self, param, value):
    value = _validate_with_unit(self, param, value)

    self._h = value.value / 100.0
    self._hubble_distance = (const.c / value).to(u.Mpc)
    H0_s = value.value * H0units_to_invs
    self._hubble_time = u.Quantity(sec_to_Gyr / H0_s, u.Gyr)

    return value
```

Similarly a lot of the auxiliary neutrino attributes could be moved into the ``m_nu.validator``.
But I think that's good for a v5.1 followup.

mhvk replied:
OK, to be honest, I prefer the simpler update-the-private attribute later too. And nothing we do here affects the API of `Parameter`, which suggests we may have converged on that!

mhvk replied:
Just to be sure, are you thinking of doing the update-the-private-attribute in this PR still? I think it has the advantage of leaving things in a consistent state, in which there is no need to use the nominally private validation functions.

nstarman replied:
Yes. The commit that I most recently pushed moved the contents of the ``m_nu`` validator into ``FLRW.__init__``. Now only ``Ob0`` has a non pre-registered validator.

mhvk replied:
Ah, now I see how it works... continuing to review...

mhvk commented on astropy/cosmology/flrw.py:
```diff
@@ -270,6 +259,39 @@ def __init__(self, H0, Om0, Ode0, Tcmb0=0.0*u.K, Neff=3.04, m_nu=0.0*u.eV,
+    # Parameter details
+
+    @m_nu.validator
+    def m_nu(self, param, _):
```
See comment above. The validator for `Ob0`, in contrast, feels just right.

mhvk commented on astropy/cosmology/parameter.py:
```diff
@@ -0,0 +1,261 @@
+
+        return register
+
+    @classproperty
```
I'd not add this - it is useful for developers only and those can be pointed to the private attribute.

nstarman replied:
Removed.

mhvk commented on docs/cosmology/dev.rst:
```diff
@@ -107,33 +106,32 @@ the definition of :class:`~astropy.cosmology.FLRW`.
-        @lazyproperty  # so only does expensive computation once
-        def m_nu(self):
-            """ Mass of neutrino species"""
+        @m_nu.setter
```
This is not correct any more, I think!

nstarman replied:
Done!

mhvk commented on docs/cosmology/dev.rst:
```diff
@@ -107,33 +106,32 @@ the definition of :class:`~astropy.cosmology.FLRW`.
 
         def __init__(self, H0, Om0, Ode0, Tcmb0=0.0*u.K, Neff=3.04, m_nu=0.0*u.eV,
                      Ob0=None, *, name=None, meta=None):
-            cls = self.__class__
```
Note: do update the parameter definitions, since these now have `fvalidate`!

nstarman replied:
Done!

mhvk commented on docs/cosmology/dev.rst:
```diff
@@ -107,33 +106,32 @@ the definition of :class:`~astropy.cosmology.FLRW`.
+``__init__``. |Parameter| allows for a value to be set once (before
+auto-locking), so ``self.H0 = H0`` will use this setter and put the value on
+"._H0". The advantage of this method over direct assignment to the private
+attribute is the use of setters / validators. |Parameter| allows for custom
```
No longer true!

nstarman replied:
Updating!

nstarman replied:
Done!

mhvk commented on docs/cosmology/dev.rst:
```diff
@@ -107,33 +106,32 @@ the definition of :class:`~astropy.cosmology.FLRW`.
+so, return the value as a |Quantity| with those units, using all enabled and
+the parameter's unit equivalencies.
+
+Note: ``Parameter.setter`` must be the top-most decorator.
```
Should talk about `validator` instead!

nstarman replied:
Done!

mhvk reviewed (commented):
Looks good, but still some missing pieces in the documentation - can they be doctested? 
Also, I'd not do the new `classproperty` - that is developer information so find to refer to a private attribute.

mhvk commented on docs/cosmology/dev.rst:
```diff
@@ -90,50 +89,49 @@ the definition of :class:`~astropy.cosmology.FLRW`.
+                return value
             ...
-            return u.Quantity(numass, self.__class__.m_nu.unit)
+            return value
```
I'd suggest to copy the whole validator, as right now it misses the point that you can access the cosmology instance easily.

mhvk commented on docs/cosmology/dev.rst:
```diff
@@ -90,50 +89,49 @@ the definition of :class:`~astropy.cosmology.FLRW`.
+auto-locking), so ``self.H0 = H0`` will use this setter and put the value on
+"._H0". The advantage of this method over direct assignment to the private
+attribute is the use of validators. |Parameter| allows for custom value
+validators, using the method-decorator ``validator``, that can check a values
```
a values -> a value's

mhvk commented on docs/cosmology/io.rst:
```diff
@@ -237,7 +237,6 @@ class to use. Details about metadata treatment are in
 ``Cosmology.from_format.help("mapping")``.
 
 .. code-block:: python
-    :emphasize-lines: 12,13
```
Seems a bit unrelated but fine to get it in too!

mhvk reviewed (approved):
This looks all great!  Just one remaining suggestion for the docs (plus a very small typo). I'll approve in the meantime.

nstarman commented:
Thanks @mhvk for the multiple rounds of review! This refactoring of Parameter is a significant improvement over the original implementation. And thanks @astrofrog for getting this in v5.0!
